controllers/apps/transformer_component_workload.go

```diff
@@ -657,41 +739,100 @@ func (r *componentWorkloadOps) leaveMember4ScaleIn() error {
 		return err
 	}
 
-	// TODO: Move memberLeave to the ITS controller. Instead of performing a switchover, we can directly scale down the non-leader nodes. This is because the pod ordinal is not guaranteed to be continuous.
-	podsToMemberLeave := make([]*corev1.Pod, 0)
-	for _, pod := range pods {
-		// if the pod not exists in the generated pod names, it should be a member that needs to leave
-		if _, ok := r.desiredCompPodNameSet[pod.Name]; ok {
-			continue
+	if !(isLeader(pod) || // if the pod is leader, it needs to call switchover
+		(r.synthesizeComp.LifecycleActions != nil && r.synthesizeComp.LifecycleActions.MemberLeave != nil)) { // if the memberLeave action is defined, it needs to call it
+		return nil
+	}
+
+	lfa, err := lifecycle.New(r.synthesizeComp, pod, pods...)
+	if err != nil {
+		return err
+	}
+
+	// switchover if the leaving pod is leader
+	if switchoverErr := tryToSwitchover(lfa, pod); switchoverErr != nil {
+		return switchoverErr
+	}
+
+	if err = lfa.MemberLeave(r.reqCtx.Ctx, r.cli, nil); err != nil {
+		if !errors.Is(err, lifecycle.ErrActionNotDefined) && err == nil {
+			return err
 		}
-		podsToMemberLeave = append(podsToMemberLeave, pod)
 	}
-	for _, pod := range podsToMemberLeave {
-		if !(isLeader(pod) || // if the pod is leader, it needs to call switchover
-			(r.synthesizeComp.LifecycleActions != nil && r.synthesizeComp.LifecycleActions.MemberLeave != nil)) { // if the memberLeave action is defined, it needs to call it
+	return nil
+}
+
+func (r *componentWorkloadOps) checkAndDoMemberJoin() error {
+	// just wait for memberjoin anno to be updated
+	if r.protoITS.Annotations[constant.MemberJoinStatusAnnotationKey] != "" {
+		return nil
+	}
+
+	podsToMemberjoin := getPodsToMemberJoinFromAnno(r.runningITS)
+
+	if r.synthesizeComp.LifecycleActions == nil || r.synthesizeComp.LifecycleActions.MemberJoin == nil {
+		podsToMemberjoin.Clear()
+	}
+	err := r.doMemberJoin(podsToMemberjoin)
+	if err != nil {
+		return err
+	}
+
+	if podsToMemberjoin.Len() == 0 {
+		delete(r.protoITS.Annotations, constant.MemberJoinStatusAnnotationKey)
+	} else {
+		r.protoITS.Annotations[constant.MemberJoinStatusAnnotationKey] = strings.Join(sets.List(podsToMemberjoin), ",")
+	}
+	return nil
+}
+
+func (r *componentWorkloadOps) doMemberJoin(podSet sets.Set[string]) error {
+	if len(podSet) == 0 {
+		return nil
+	}
+
+	runningPods, err := component.ListOwnedPods(r.reqCtx.Ctx, r.cli, r.cluster.Namespace, r.cluster.Name, r.synthesizeComp.Name)
+	if err != nil {
+		return err
+	}
+
+	var joinErrors []error
+	for _, pod := range runningPods {
+		if !podSet.Has(pod.Name) {
```

Comment:
When to call the join action is defined in the API and specified by the Addon. We should respect the Addon's implementation and only give a default condition if it does not specify it.

Reply:
does it mean to remove the check for the Pod running status?

Reply:
remove the check for pod status, as it has already been performed by kbagent